Grafana configuration requires explicit port number broken in OpenShift

`GET /health` sees this in logs:

```
[ERROR] io.vertx.ext.web.handler.impl.HttpStatusException: Internal Server Error
Caused by: java.lang.IllegalArgumentException: port p must be in range 0 <= p <= 65535
	at io.vertx.core.impl.Arguments.requireInRange(Arguments.java:44)
	at io.vertx.core.net.impl.SocketAddressImpl.<init>(SocketAddressImpl.java:32)
	at io.vertx.core.net.SocketAddress.inetSocketAddress(SocketAddress.java:37)
	at io.vertx.core.http.impl.HttpClientImpl.createRequest(HttpClientImpl.java:1106)
	at io.vertx.core.http.impl.HttpClientImpl.createRequest(HttpClientImpl.java:1074)
	at io.vertx.core.http.impl.HttpClientImpl.request(HttpClientImpl.java:591)
	at io.vertx.ext.web.client.impl.HttpContext.handlePrepareRequest(HttpContext.java:344)
	at io.vertx.ext.web.client.impl.HttpContext.execute(HttpContext.java:288)
	at io.vertx.ext.web.client.impl.HttpContext.next(HttpContext.java:275)
	at io.vertx.ext.web.client.impl.predicate.PredicateInterceptor.handle(PredicateInterceptor.java:69)
	at io.vertx.ext.web.client.impl.predicate.PredicateInterceptor.handle(PredicateInterceptor.java:32)
	at io.vertx.ext.web.client.impl.HttpContext.next(HttpContext.java:272)
	at io.vertx.ext.web.client.impl.HttpContext.fire(HttpContext.java:282)
	at io.vertx.ext.web.client.impl.HttpContext.prepareRequest(HttpContext.java:168)
	at io.vertx.ext.web.client.impl.HttpRequestImpl.send(HttpRequestImpl.java:331)
	at io.vertx.ext.web.client.impl.HttpRequestImpl.send(HttpRequestImpl.java:301)
	at com.redhat.rhjmc.containerjfr.net.web.handlers.HealthGetHandler.checkUri(HealthGetHandler.java:129)
	at com.redhat.rhjmc.containerjfr.net.web.handlers.HealthGetHandler.handle(HealthGetHandler.java:102)
	at com.redhat.rhjmc.containerjfr.net.web.handlers.HealthGetHandler.handle(HealthGetHandler.java:63)
	at io.vertx.ext.web.impl.BlockingHandlerDecorator.lambda$handle$0(BlockingHandlerDecorator.java:48)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)
	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

`oc describe` snippet:

```
    Environment:
      CONTAINER_JFR_PLATFORM:             com.redhat.rhjmc.containerjfr.platform.openshift.OpenShiftPlatformStrategy
      CONTAINER_JFR_SSL_PROXIED:          true
      CONTAINER_JFR_ALLOW_UNTRUSTED_SSL:  true
      CONTAINER_JFR_WEB_PORT:             8181
      CONTAINER_JFR_EXT_WEB_PORT:         443
      CONTAINER_JFR_WEB_HOST:             containerjfr-default.apps-crc.testing
      CONTAINER_JFR_LISTEN_PORT:          9090
      CONTAINER_JFR_EXT_LISTEN_PORT:      443
      CONTAINER_JFR_LISTEN_HOST:          containerjfr-command-default.apps-crc.testing
      GRAFANA_DASHBOARD_URL:              https://containerjfr-grafana-default.apps-crc.testing
      GRAFANA_DATASOURCE_URL:             http://172.25.131.63:8080
      CONTAINER_JFR_DISABLE_JMX_AUTH:     true
      CONTAINER_JFR_DISABLE_SSL:          true
```

The problematic execution pointed to by the stack trace appears to be (`HealthGetHandler.java`):

```
        checkUri(GRAFANA_DASHBOARD_ENV, "/api/health", dashboardAvailable);

    private void checkUri(String envName, String path, CompletableFuture<Boolean> future) {
        if (this.env.hasEnv(envName)) {
            URI uri;
            try {
                uri = new URI(this.env.getEnv(envName));
            } catch (URISyntaxException e) {
                logger.warn(e);
                future.complete(false);
                return;
            }
            webClient
                    .get(uri.getPort(), uri.getHost(), path)
                    .ssl("https".equals(uri.getScheme()))
                    .timeout(5000)
                    .send(
                            handler -> {
                                if (handler.failed()) {
                                    future.complete(false);
                                    this.logger.info(new IOException(handler.cause()));
                                    return;
                                }
                                future.complete(handler.result().statusCode() == 200);
                            });
        } else {
            future.complete(false);
        }
    }

```

I think, from a reading, that the problem is that the `uri` is `https://containerjfr-grafana-default.apps-crc.testing`, which has no explicit port number, so `uri.getPort()` is `-1` and the request send fails.

--- a/src/main/java/com/redhat/rhjmc/containerjfr/net/NetworkModule.java
+++ b/src/main/java/com/redhat/rhjmc/containerjfr/net/NetworkModule.java
@@ -113,8 +113,9 @@ public abstract class NetworkModule {
                     new WebClientOptions()
                             .setSsl(true)
                             .setDefaultHost(netConf.getWebServerHost())
-                            .setDefaultPort(netConf.getExternalWebServerPort());
-            if (!netConf.isUntrustedSslAllowed()) {
+                            .setDefaultPort(netConf.getExternalWebServerPort())
+                            .setFollowRedirects(true);
+            if (netConf.isUntrustedSslAllowed()) {
                 opts = opts.setTrustAll(true).setVerifyHost(false);
             }
             return WebClient.create(vertx, opts);

--- a/src/main/java/com/redhat/rhjmc/containerjfr/net/web/handlers/HealthGetHandler.java
+++ b/src/main/java/com/redhat/rhjmc/containerjfr/net/web/handlers/HealthGetHandler.java
@@ -120,10 +120,11 @@ class HealthGetHandler implements RequestHandler {
             try {
                 uri = new URI(this.env.getEnv(envName));
             } catch (URISyntaxException e) {
-                logger.warn(e);
+                logger.error(e);
                 future.complete(false);
                 return;
             }
+            logger.debug(String.format("Testing health of %s=%s %s", envName, uri.toString(), path));
             HttpRequest<Buffer> req = webClient.get(uri.getHost(), path);
             if (uri.getPort() != -1) {
                 req = req.port(uri.getPort());
@@ -133,8 +134,8 @@ class HealthGetHandler implements RequestHandler {
                     .send(
                             handler -> {
                                 if (handler.failed()) {
+                                    this.logger.warn(new IOException(handler.cause()));
                                     future.complete(false);
-                                    this.logger.info(new IOException(handler.cause()));
                                     return;
                                 }
                                 future.complete(handler.result().statusCode() == 200);

--- a/src/main/java/com/redhat/rhjmc/containerjfr/net/web/handlers/HealthGetHandler.java
+++ b/src/main/java/com/redhat/rhjmc/containerjfr/net/web/handlers/HealthGetHandler.java
@@ -124,7 +124,8 @@ class HealthGetHandler implements RequestHandler {
                 future.complete(false);
                 return;
             }
-            logger.debug(String.format("Testing health of %s=%s %s", envName, uri.toString(), path));
+            logger.debug(
+                    String.format("Testing health of %s=%s %s", envName, uri.toString(), path));
             HttpRequest<Buffer> req = webClient.get(uri.getHost(), path);
             if (uri.getPort() != -1) {
                 req = req.port(uri.getPort());

--- a/src/main/java/com/redhat/rhjmc/containerjfr/net/web/handlers/HealthGetHandler.java
+++ b/src/main/java/com/redhat/rhjmc/containerjfr/net/web/handlers/HealthGetHandler.java
@@ -55,9 +55,11 @@ import com.redhat.rhjmc.containerjfr.core.log.Logger;
 import com.redhat.rhjmc.containerjfr.core.sys.Environment;
 import com.redhat.rhjmc.containerjfr.net.web.HttpMimeType;
 
+import io.vertx.core.buffer.Buffer;
 import io.vertx.core.http.HttpHeaders;
 import io.vertx.core.http.HttpMethod;
 import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.client.HttpRequest;
 import io.vertx.ext.web.client.WebClient;
 
 class HealthGetHandler implements RequestHandler {
@@ -122,8 +124,13 @@ class HealthGetHandler implements RequestHandler {
                 future.complete(false);
                 return;
             }
-            webClient
-                    .get(uri.getPort(), uri.getHost(), path)
+            HttpRequest<Buffer> req;
+            if (uri.getPort() != -1) {
+                req = webClient.get(uri.getPort(), uri.getHost(), path);
+            } else {
+                req = webClient.get(uri.getHost(), path);
+            }
+            req
                     .ssl("https".equals(uri.getScheme()))
                     .timeout(5000)
                     .send(

--- a/src/main/java/com/redhat/rhjmc/containerjfr/net/web/handlers/HealthGetHandler.java
+++ b/src/main/java/com/redhat/rhjmc/containerjfr/net/web/handlers/HealthGetHandler.java
@@ -130,8 +130,7 @@ class HealthGetHandler implements RequestHandler {
             } else {
                 req = webClient.get(uri.getHost(), path);
             }
-            req
-                    .ssl("https".equals(uri.getScheme()))
+            req.ssl("https".equals(uri.getScheme()))
                     .timeout(5000)
                     .send(
                             handler -> {

--- a/src/main/java/com/redhat/rhjmc/containerjfr/net/web/handlers/HealthGetHandler.java
+++ b/src/main/java/com/redhat/rhjmc/containerjfr/net/web/handlers/HealthGetHandler.java
@@ -124,11 +124,9 @@ class HealthGetHandler implements RequestHandler {
                 future.complete(false);
                 return;
             }
-            HttpRequest<Buffer> req;
+            HttpRequest<Buffer> req = webClient.get(uri.getHost(), path);
             if (uri.getPort() != -1) {
-                req = webClient.get(uri.getPort(), uri.getHost(), path);
-            } else {
-                req = webClient.get(uri.getHost(), path);
+                req = req.port(uri.getPort());
             }
             req.ssl("https".equals(uri.getScheme()))
                     .timeout(5000)

--- a/src/test/java/com/redhat/rhjmc/containerjfr/net/web/handlers/HealthGetHandlerTest.java
+++ b/src/test/java/com/redhat/rhjmc/containerjfr/net/web/handlers/HealthGetHandlerTest.java
@@ -223,4 +223,52 @@ class HealthGetHandlerTest {
         Assertions.assertEquals(responseMap.get("dashboardAvailable"), true);
         Assertions.assertEquals(responseMap.get("datasourceAvailable"), false);
     }
+
+    @Test
+    void shouldHandleHealthRequestWithDashboardUrlWithoutExplicitPort() {
+        RoutingContext ctx = mock(RoutingContext.class);
+        HttpServerResponse rep = mock(HttpServerResponse.class);
+        when(ctx.response()).thenReturn(rep);
+        when(rep.putHeader(Mockito.any(CharSequence.class), Mockito.anyString())).thenReturn(rep);
+
+        String url = "https://hostname/";
+        when(env.hasEnv("GRAFANA_DASHBOARD_URL")).thenReturn(true);
+        when(env.getEnv("GRAFANA_DASHBOARD_URL")).thenReturn(url);
+        when(env.hasEnv("GRAFANA_DATASOURCE_URL")).thenReturn(false);
+
+        HttpRequest<Buffer> req = Mockito.mock(HttpRequest.class);
+        HttpResponse<Buffer> resp = Mockito.mock(HttpResponse.class);
+        Mockito.when(webClient.get(Mockito.anyString(), Mockito.anyString()))
+                .thenReturn(req);
+        Mockito.when(req.ssl(Mockito.anyBoolean())).thenReturn(req);
+        Mockito.when(req.timeout(Mockito.anyLong())).thenReturn(req);
+        Mockito.doAnswer(
+                        new Answer<Void>() {
+                            @Override
+                            public Void answer(InvocationOnMock args) throws Throwable {
+                                AsyncResult<HttpResponse<Buffer>> asyncResult =
+                                        Mockito.mock(AsyncResult.class);
+                                Mockito.when(asyncResult.result()).thenReturn(resp);
+                                Mockito.when(resp.statusCode()).thenReturn(200);
+                                ((Handler<AsyncResult<HttpResponse<Buffer>>>) args.getArgument(0))
+                                        .handle(asyncResult);
+                                return null;
+                            }
+                        })
+                .when(req)
+                .send(Mockito.any());
+
+        handler.handle(ctx);
+
+        verify(rep).putHeader(HttpHeaders.CONTENT_TYPE, HttpMimeType.JSON.mime());
+        ArgumentCaptor<String> responseCaptor = ArgumentCaptor.forClass(String.class);
+        verify(rep).end(responseCaptor.capture());
+
+        Map<String, Boolean> responseMap =
+                gson.fromJson(
+                        responseCaptor.getValue(),
+                        new TypeToken<Map<String, Boolean>>() {}.getType());
+        Assertions.assertEquals(responseMap.get("dashboardAvailable"), true);
+        Assertions.assertEquals(responseMap.get("datasourceAvailable"), false);
+    }
 }

--- a/src/test/java/com/redhat/rhjmc/containerjfr/net/web/handlers/HealthGetHandlerTest.java
+++ b/src/test/java/com/redhat/rhjmc/containerjfr/net/web/handlers/HealthGetHandlerTest.java
@@ -142,8 +142,8 @@ class HealthGetHandlerTest {
 
         HttpRequest<Buffer> req = Mockito.mock(HttpRequest.class);
         HttpResponse<Buffer> resp = Mockito.mock(HttpResponse.class);
-        Mockito.when(webClient.get(Mockito.anyInt(), Mockito.anyString(), Mockito.anyString()))
-                .thenReturn(req);
+        Mockito.when(webClient.get(Mockito.anyString(), Mockito.anyString())).thenReturn(req);
+        Mockito.when(req.port(Mockito.anyInt())).thenReturn(req);
         Mockito.when(req.ssl(Mockito.anyBoolean())).thenReturn(req);
         Mockito.when(req.timeout(Mockito.anyLong())).thenReturn(req);
         Mockito.doAnswer(
@@ -190,8 +190,8 @@ class HealthGetHandlerTest {
 
         HttpRequest<Buffer> req = Mockito.mock(HttpRequest.class);
         HttpResponse<Buffer> resp = Mockito.mock(HttpResponse.class);
-        Mockito.when(webClient.get(Mockito.anyInt(), Mockito.anyString(), Mockito.anyString()))
-                .thenReturn(req);
+        Mockito.when(webClient.get(Mockito.anyString(), Mockito.anyString())).thenReturn(req);
+        Mockito.when(req.port(Mockito.anyInt())).thenReturn(req);
         Mockito.when(req.ssl(Mockito.anyBoolean())).thenReturn(req);
         Mockito.when(req.timeout(Mockito.anyLong())).thenReturn(req);
         Mockito.doAnswer(

--- a/src/test/java/com/redhat/rhjmc/containerjfr/net/web/handlers/HealthGetHandlerTest.java
+++ b/src/test/java/com/redhat/rhjmc/containerjfr/net/web/handlers/HealthGetHandlerTest.java
@@ -238,8 +238,7 @@ class HealthGetHandlerTest {
 
         HttpRequest<Buffer> req = Mockito.mock(HttpRequest.class);
         HttpResponse<Buffer> resp = Mockito.mock(HttpResponse.class);
-        Mockito.when(webClient.get(Mockito.anyString(), Mockito.anyString()))
-                .thenReturn(req);
+        Mockito.when(webClient.get(Mockito.anyString(), Mockito.anyString())).thenReturn(req);
         Mockito.when(req.ssl(Mockito.anyBoolean())).thenReturn(req);
         Mockito.when(req.timeout(Mockito.anyLong())).thenReturn(req);
         Mockito.doAnswer(

